currency/pairs.go

```diff
@@ -199,28 +199,25 @@ func (p Pairs) GetPairsByCurrencies(currencies Currencies) Pairs {
 	return pairs
 }
 
-// Remove removes the specified pair from the list of pairs if it exists
-func (p Pairs) Remove(pair Pair) (Pairs, error) {
-	pairs := slices.Clone(p)
-	for x := range p {
-		if p[x].Equal(pair) {
-			return append(pairs[:x], pairs[x+1:]...), nil
+// Remove removes the specified pairs from the list of pairs if they exist
+func (p Pairs) Remove(rem ...Pair) Pairs {
+	var n Pairs
+	for _, pN := range p {
+		if !slices.Contains(rem, pN) {
+			n = append(n, pN)
 		}
 	}
-	return nil, fmt.Errorf("%s %w", pair, ErrPairNotFound)
+	return n
 }
 
-// Add adds a specified pair to the list of pairs if it doesn't exist
+// Add adds pairs to the list of pairs ignoring duplicates
 func (p Pairs) Add(pairs ...Pair) Pairs {
```

Comment:
Looking at benchmark this is indeed more performant but if p Pairs is pre allocated with a capacity greater that the incoming slice then you are modifiying the original slice. It's an edge case that was included in the previous tests.

Reply:
I don't see evidence for this edge case.
Yes, the previous test proved that it wasn't modified.
But that doesn't prove it would be modified otherwise.
I added a test for this, used slices.Grow, and then removed the slices.Clone (or Clip) and my test still passed.
I forced the issue by changing the original to be sure.
Pass-by-value and then append does not modify the original.

HOWEVER: If you're meaning that we'll *return* a slice where modifying it would change the original, then agreed. But the existing tests don't check for that, right?

Anyway, I have tests and a Clone to cover all bases now.

Reply:
Yeah it's a bit of a weird one and hard to see I should have gotten @cranktakular to break it out a bit better. It was central to this block:

```go
	// Test whether the original pairs have been modified
	pairsWithExtraBaggage := make(Pairs, 0, len(p)+3)
	pairsWithExtraBaggage = append(pairsWithExtraBaggage, p...)
	brain := NewPair(BRAIN, USD)
	// Edge case: using preallocated pairsWithExtraBaggage as base for two independent slices
	// will leak changes across underlying arrays.
	withBrainPairs := pairsWithExtraBaggage.Add(NewPair(BTC, LTC), brain)
	fmt.Println("withBrainPairs", withBrainPairs)
	assert.Equal(t, brain, withBrainPairs[len(withBrainPairs)-1])
	badger := NewPair(BADGER, USD)
	withBadgerPairs := pairsWithExtraBaggage.Add(NewPair(BTC, LTC), badger)
	fmt.Println("withBadgerPairs", withBadgerPairs)
	fmt.Println("withBrainPairs second check", withBrainPairs)  // This will be modified with magic and contain badger
	fmt.Println("pairsWithExtraBaggage", pairsWithExtraBaggage) // This won't be modified, so its hard to test.
	assert.Equal(t, badger, withBadgerPairs[len(withBadgerPairs)-1])
	assert.Equal(t, brain, withBrainPairs[len(withBrainPairs)-1])
```

Adding this block and reverting the changes should pass the tests again.

Reply:
Fixed 7604877

Reply:
Also Fixed cc1e73b

All our issues are symptoms of underlying array not re-allocated.
But we'd also fall into an issue of changes within the arrays contaminating across siblings.
So the fix and the test is just to Clone() it and ensure it doesn't cross-contaminate.